[TASK] 7.2 compatibility (3.1.x)

AndreasA opened the pull request.

irnnr commented on Classes/Backend/IndexingConfigurationSelectorField.php:
```diff
@@ -102,12 +102,15 @@ public function getSelectedValues() {
-		$tceForm = t3lib_div::makeInstance('t3lib_TCEforms');
+		$selectFieldRenderer = $formEngine = \TYPO3\CMS\Core\Utility\GeneralUtility::makeInstance('TYPO3\\CMS\\Backend\\Form\\FormEngine');
+		if (class_exists('TYPO3\\CMS\Backend\\Form\\Element\\SelectElement')) {
+			$selectFieldRenderer = \TYPO3\CMS\Core\Utility\GeneralUtility::makeInstance('TYPO3\\CMS\Backend\\Form\\Element\\SelectElement', $formEngine);
```
When would this be the case? / When would SelectElement not be available?

AndreasA replied:
SelectElement is new for 7.x
With this check it works also for 6.2 because there the SelectElement does not exist and it would use FormEngine to render as before.
I could also add a version check but I personally prefer feature checks. Maybe we could also check if formengine doesn't have the required render method?

irnnr commented on Classes/PluginBase/PluginBase.php:
```diff
@@ -160,11 +160,10 @@ protected function baseWrap($content) {
-			$GLOBALS['TSFE']->tmpl->setup['plugin.']['tx_solr.'],
-			$this->conf
+		$this->conf = $GLOBALS['TSFE']->tmpl->setup['plugin.']['tx_solr.'];
+		\TYPO3\CMS\Core\Utility\ArrayUtility::mergeRecursiveWithOverrule(
```
why the order change? Also for the next chunk.

AndreasA replied:
Because mergeRecursiveWithOverrule is now passed by reference.
Before it returned the result.
Which means if I want the same result as before, I have $this->conf be $GLOBALS['TSFE']->tmpl->setup['plugin.']['tx_solr.']

otherwise $GLOBALS['TSFE']->tmpl->setup['plugin.']['tx_solr.'] would get the result.

irnnr replied:
Is this new in 7.x?

irnnr commented on EidSuggest/Suggest.php:
```diff
@@ -25,43 +25,37 @@
-$TSFE->rootLine = $TSFE->sys_page->getRootLine($pageId, '');
-$TSFE->initTemplate();
-$TSFE->getConfigArray();
+\TYPO3\CMS\Frontend\Utility\EidUtility::initTCA();
```
Is this compatible with 6.2?

AndreasA replied:
Yes initTca is 6.2 compatible:
https://git.typo3.org/Packages/TYPO3.CMS.git/blob/f695f377f48c23bec3156c6d0f58f8a078d8a651:/typo3/sysext/frontend/Classes/Utility/EidUtility.php#l78

The $TSFE is not exactly removed but was replaced with $GLOBALS['TSFE'] instead.

irnnr commented on EidSuggest/Suggest.php:
```diff
@@ -25,43 +25,37 @@
+$pageId     = filter_var(\TYPO3\CMS\Core\Utility\GeneralUtility::_GET('id'), FILTER_SANITIZE_NUMBER_INT);
 $languageId = filter_var(
-	t3lib_div::_GET('L'),
+	\TYPO3\CMS\Core\Utility\GeneralUtility::_GET('L'),
```
Please import the classes

AndreasA replied:
Cannot use "use" here as it is and eID script not a class definition.
Or at least it was always avoided in core if it is only a file.

irnnr replied:
That should not matter

AndreasA replied:
True you can import without class too, it is just not generally done in the core.
But I guess that's mostly because most of the times it is required in ext_* or TCA and those are merged during caching.
But in the dumpFile eID it isn't done with imports either but that could just be because of conversion from old classnames to new ones.

irnnr commented on Resources/Private/Layouts/Default.html:
```diff
@@ -1,7 +1,7 @@
 
 <f:be.container
-	addCssFile="{f:uri.resource(path:'StyleSheets/ModuleAdministration.css')}"
+	includeCssFiles="{solrAdmin: '{f:uri.resource(path:\'StyleSheets/ModuleAdministration.css\')}'}"
```
6.2 compatibility?

AndreasA replied:
addCssFile works in 6.2 too but was deprecated.
Could also go back there...

irnnr replied:
ok, thx

irnnr commented:
Overall good PR!

AndreasA commented:
Added corresponding inline comments.
Let me know if I have to change anything. Especially, regarding SelectElement.

Added another change (setup.txt has the wrong class name for PageIndexer (lowercase instead of corresponding uppercase) which might make issue with composer class loading in 7.3

AndreasA commented:
Same for initialize class of page indexer

AndreasA commented:
Added change to handle the select field provider in 7.3 / 7.2 and 6.2
Maybe you can verify that it also works in older versions.

Not sure, however, if this is the best solution. Maybe it would be better to extend the SelectCheckboxElement instead and provide a mapper for the now protected method. Of course, that would only be possible if the corresponding class exists.

AndreasA commented:
another wrong class which results in errors in 7.3 (due to wrong case used in class name)

AndreasA commented:
Does this still need changes? If so, which?

irnnr commented:
I'll review after we have 3.0.1 out, thanks!

AndreasA commented:
No problem, take your time to review it.
Especially to ensure it works with also still works correctly with 6.2.
It should but I mostly tested it with later versions.

irnnr commented:
LGTM, just needs a rebase
